### PR TITLE
[PIR][DynamicShape] Fix Gather Op and Shape Op && Add BC_binary Ops' inferSymbolic shape

### PR DESCRIPTION
**Pull request by zhangboSJTU**
### PR types

Others

### PR changes

Others

### Description

Pcard-67164

- Add all broadcast_binary Ops' inferSymbolic shape
- Sort the funcs by Op name alphabetically
- add print op's inferSymbolic shape
- fix Gather Op's and Shape Op's inferSymbolic shape

### CODE REVIEW

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/paddle_op_infer_sym.cc [hunk 1]
```diff
@@ -59,20 +59,12 @@ bool ShapeOpInferSymbolicShape(pir::Operation *op,
...
-    std::vector<symbol::DimExpr> sym_shape;
-    symbol::DimExpr dim_expr(
-        op->result(0).type().dyn_cast<pir::DenseTensorType>().dims()[0]);
-    sym_shape.emplace_back(dim_expr);
-    return sym_shape;
-  }();
-
-  symbol::ShapeOrDataDimExprs shape_or_data{symbol::TensorShapeOrDataDimExprs(
-      sym_shape, operand_shape_or_data.shape())};
+  const std::vector<symbol::DimExpr> out_data = operand_shape_or_data.shape();
```
Comment (tc20042008) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/paddle_op_infer_sym.cc]: const auto&

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/paddle_op_infer_sym.cc [hunk 2]
```diff
@@ -1234,22 +1202,65 @@ bool DirichletOpInferSymbolicShape(
...
+  const std::vector<symbol::DimExpr> &index_sym_shape =
+      index_shape_or_data.data().has_value()
+          ? index_shape_or_data.data().value()
+          : index_shape_or_data.shape();
+
+  int axis =
+      static_cast<int>(axis_shape_or_data.data().value()[0].Get<int64_t>());
+  if (axis < 0) axis += input_sym_shape.size();
+
+  std::vector<symbol::DimExpr> out_sym_shape;
```
Comment (tc20042008) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/paddle_op_infer_sym.cc]: 无脑地使用lambda init初始化`out_sym_shape`。
原因是上面的input_sym_shape都是为了out_sym_shape服务的，前者都应该放到后者的初始化lambda里

Comment (zhangboSJTU) [reply paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/paddle_op_infer_sym.cc]: done

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/paddle_op_infer_sym.cc (hunk 1 above)
Comment (zhangboSJTU) [reply paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/paddle_op_infer_sym.cc]: thanks~

**Review by tc20042008 — DISMISSED** [review]

**Review by zyfncg — APPROVED** [review]

**Pull request merged**